pkg/urbit/c/bile.c

```diff
@@ -0,0 +1,162 @@
+//! @file bile.c
+
+#include "c/bile.h"
+
+#include <arpa/inet.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <stddef.h>
+#include <sys/stat.h>
+
+#include "c/defs.h"
+
+//==============================================================================
+// Macros
+//==============================================================================
+
+//! Seek to the beginning of a binary file.
+#define seek_to_beginning(bile_handle, failure_action)                         \
+  do {                                                                         \
+    try_syscall(lseek((bile_handle)->fid_i, 0, SEEK_SET),                      \
+                failure_action,                                                \
+                "failed to seek to start of %s",                               \
+                (bile_handle)->pax_c);                                         \
+  } while ( 0 )
+
+//==============================================================================
+// Functions
+//==============================================================================
+
+c3_bile*
+c3_bile_open(const c3_c* const pax_c)
+{
+  if ( !pax_c ) {
+    goto fail;
+  }
+
+  c3_bile* bil_u = c3_calloc(sizeof(*bil_u));
+
+  bil_u->pax_c = c3_calloc(1 + strlen(pax_c));
+  strcpy(bil_u->pax_c, pax_c);
+
+  try_syscall(bil_u->fid_i = open(bil_u->pax_c, O_RDWR | O_CREAT, 0666),
+              goto free_file,
+              "failed to open %s",
+              bil_u->pax_c);
+
+  struct stat buf_u;
+  try_syscall(fstat(bil_u->fid_i, &buf_u),
+              goto close_file,
+              "failed to get %s stats",
+              bil_u->pax_c);
+
+  bil_u->len_w = buf_u.st_size;
+  goto succeed;
+
+close_file:
+  close(bil_u->fid_i);
+free_file:
+  c3_free(bil_u->pax_c);
+  c3_free(bil_u);
+fail:
+  return NULL;
+
+succeed:
+  return bil_u;
+}
+
+//! @n (1) Overwrite any existing data.
+//! @n (2) Attempt the write again if we were interrupted by a signal.
+//! @n (3) Chop off any data that remains from a previous call.
+//! @n (4) Writing failed partway through, so clear the file.
```

Comment:
What's the advantage of placing these comments here and associating them with target code via indexing instead of just placing them over the target code?

Reply:
See the explanation in https://github.com/urbit/urbit/pull/5602#discussion_r824237092.